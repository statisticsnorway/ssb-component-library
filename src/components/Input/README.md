Input
========

> Input field and search field

### Usage

#### HTML

```html
<div class="ssb-input">
  <label for="input1">Insert label</label>
  <div class="input-wrapper">
    <input id="input1" value="option1" />
  </div>
</div>

<div class="ssb-input">
  <label for="input1">Insert label</label>
  <div class="input-wrapper">
    <input id="input1" value="option1" />
    <div class="icon-wrapper">
      <icon>search</icon>
    </div>
  </div>
</div>

<div class="ssb-input error">
  <label for="input1">Insert label</label>
  <div class="input-wrapper">
    <input id="input1" value="option1" />
    <div class="icon-wrapper">
      <icon>alert</icon>
    </div>
  </div>
  <span class="ssb-form-error">Insert error message</span>
</div>
```

#### React

```jsx harmony
<Input
  handleChange={handleChange}
  value={someValue}
/>

<Input
  searchField
  submitCallback={handleSubmit}
  value={someValue}
/>

<Input
  error
  errorMessage="Dette er feil"
/>
```

Available props:

| Name       | Type           | Description  |
| ---------- | ------------- | ----- |
| disabled | bool | Toggles disabled state |
| error | bool | Changes design and inserts error icon |
| errorMessage | string | Error message to be displayed |
| handleChange | func | Callback every time a letter is entered or removed |
| id | string | ID to connect label and input field |
| label | string | Label text |
| negative | bool | Changes design |
| searchField | bool | Inserts a search icon |
| submitCallback | func | Callback for when clicking search icon |
| type | string | Optional parameters like 'url', 'email' etc. |
| value | string | Current value |
