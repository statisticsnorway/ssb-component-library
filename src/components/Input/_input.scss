.ssb-input {
  @include roboto;
  cursor: text;
  display: flex;
  flex-direction: column;
  max-width: 400px;
  position: relative;
  width: 100%;

  &:hover {

    input {
      border: 1px solid $ssb-green-4;
      outline: 1px solid $ssb-green-4;

      &:disabled {
        border: 1px solid $ssb-dark-5;
      }
    }
  }

  .input-wrapper {
    display: flex;
  }

  label {
    color: $ssb-dark-6;
    cursor: text;
    font-size: 14px;
    margin-bottom: 4px;
    user-select: none;
  }

  input {
    @include roboto;
    background: $ssb-white;
    border: 1px solid $ssb-dark-5;
    box-sizing: border-box;
    color: $ssb-dark-6;
    font-size: 16px;
    height: $input-field-height;
    padding: 12px 12px;
    text-overflow: ellipsis;
    width: 100%;

    &.with-icon {
      padding: 4px $input-field-height 4px 10px;
    }

    &:disabled {
      border: 1px solid $ssb-dark-5;
      cursor: not-allowed;
    }

    &:focus {
      border: 1px solid $ssb-green-4;
      outline: 1px solid $ssb-green-4;
      outline-offset: 0;
    }
  }

  &.negative {

    .input-wrapper {

      ::placeholder {
        color: $ssb-white;
        opacity: 1;
      }

      :-ms-input-placeholder {
        color: $ssb-white;
      }

      ::-ms-input-placeholder {
        color: $ssb-white;
      }
    }

    &:hover {

      input {
        border: 1px solid $ssb-green-2;
        outline: 1px solid $ssb-green-2;

        &:disabled {
          border: 1px solid $ssb-white;
        }
      }
    }

    label {
      color: $ssb-white;
    }

    input {
      background: transparent;
      border: 1px solid $ssb-white;
      color: $ssb-white;

      &:disabled {
        border: 1px solid $ssb-white;
      }

      &:focus {
        border: 1px solid $ssb-green-2;
        outline: 1px solid $ssb-green-2;
        outline-offset: 0;
      }
    }

    .icon-wrapper {

      & > * {
        color: $ssb-green-2;
      }

      &.search-icon {
        cursor: pointer;

        &:hover, &:focus {
          background: $ssb-white;

          & > * {
            color: $ssb-dark-5;
          }
        }
      }
    }

    &.error {

      input {
        border: 1px solid $ssb-red-3;
        outline: 1px solid $ssb-red-3;
      }

      .icon-wrapper {

        & > * {
          color: $ssb-red-3;
        }
      }
    }
  }

  &.error {

    input {
      border: 1px solid $ssb-red-3;
      outline: 1px solid $ssb-red-3;
    }

    .icon-wrapper {

      & > * {
        color: $ssb-red-3;
      }
    }
  }

  .icon-wrapper {
    @include focus-marker;
    @include reset-button;
    align-items: center;
    display: flex;
    font-size: 18px;
    height: $input-field-height;
    justify-content: center;
    margin-left: -$input-field-height;
    width: $input-field-height;

    & > * {
      color: $ssb-green-4;
    }

    &.search-icon {
      cursor: pointer;

      &:hover, &:focus {
        background: $ssb-green-4;

        & > * {
          color: $ssb-white;
        }
      }
    }
  }
}
