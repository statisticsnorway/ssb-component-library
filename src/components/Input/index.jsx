import React, { useState } from 'react';
import PropTypes from 'prop-types';
import uuid from 'uuid/v4';
import { AlertTriangle, Search }Â from 'react-feather';
import FormError from '../FormError/index.jsx';

const Input = ({
	disabled, error, errorMessage, handleChange, id, label, negative, searchField, submitCallback, type, value,
}) => {
	const [inputValue, setValue] = useState(value);
	const inputId = id || uuid();
	const handleInputChange = e => {
		setValue(e.target.value);
		handleChange(e.target.value);
	};

	return (
		<div className={`ssb-input ${negative ? 'negative' : ''} ${error ? 'error' : ''}`}>
			<label htmlFor={inputId}>{label}</label>
			<div className="input-wrapper">
				<input
					id={inputId}
					disabled={disabled}
					type={type}
					value={inputValue}
					onChange={e => handleInputChange(e)}
				/>
				{searchField && (
					<div className="icon-wrapper" onClick={() => submitCallback(inputValue)}>
						<Search size={18} />
					</div>
				)}
				{error && (
					<div className="icon-wrapper">
						<AlertTriangle size={18} />
					</div>
				)}
			</div>
			{error && (errorMessage && (
				<FormError negative={negative} errorMessage={errorMessage} />
			))}
		</div>
	);
};

Input.defaultProps = {
	disabled: false,
	error: false,
	handleChange: () => {},
	label: 'Input field',
	negative: false,
	searchField: false,
	submitCallback: () => {},
	type: 'text',
};

Input.propTypes = {
	disabled: PropTypes.bool,
	error: PropTypes.bool,
	errorMessage: PropTypes.string,
	handleChange: PropTypes.func,
	id: PropTypes.string,
	label: PropTypes.string,
	negative: PropTypes.bool,
	searchField: PropTypes.bool,
	submitCallback: PropTypes.func,
	type: PropTypes.string,
	value: PropTypes.string,
};

export default Input;
