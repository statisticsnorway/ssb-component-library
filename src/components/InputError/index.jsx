import React from 'react';
import PropTypes from 'prop-types';

const InputError = ({ errorMessage, negative }) => (
	<span className={`ssb-input-error ${negative ? 'negative' : ''}`}>{errorMessage}</span>
);

InputError.defaultProps = {
	negative: false,
};

InputError.propTypes = {
	errorMessage: PropTypes.string.isRequired,
	negative: PropTypes.bool,
};

export default InputError;
