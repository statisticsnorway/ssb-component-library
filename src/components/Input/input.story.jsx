import React from 'react'
import Input from './index'

let someValue = ''
const handleChange = (e) => {
  someValue = e
}

const handleSubmit = (e) => {
  console.log(e)
}

export default {
  title: 'Input',
}

export const Default = () => (
  <div style={{ width: '280px' }}>
    <Input label='Input field' value={someValue} handleChange={handleChange} />
  </div>
)

export const SearchField = () => (
  <div style={{ width: '280px' }}>
    <Input
      ariaLabel='Input field Search'
      ariaLabelSearchButton='Search'
      searchField
      submitCallback={handleSubmit}
      placeholder='Search text'
    />
  </div>
)

export const WithValue = () => (
  <div style={{ width: '280px' }}>
    <Input label='Input field' value='Already filled' />
  </div>
)

export const Disabled = () => (
  <div style={{ width: '280px' }}>
    <Input label='Input field' disabled />
  </div>
)

export const Error = () => (
  <div style={{ width: '280px' }}>
    <Input label='Input field' error errorMessage='Beklager, dette er feil' />
  </div>
)

export const Negative = () => (
  <div
    style={{
      width: '40em',
      height: '20em',
      background: '#274247',
      display: 'flex',
      flexDirection: 'column',
      justifyContent: 'center',
      alignItems: 'center',
    }}
  >
    <Input label='Input field' negative value={someValue} handleChange={handleChange} />
    <Input label='Input field' negative searchField submitCallback={handleSubmit} placeholder='Search text' />
    <Input label='Input field' negative value={someValue} error errorMessage='Beklager, dette er feil' />
  </div>
)

export const LargeInput = () => (
  <div
    style={{
      width: '980px',
      height: '240px',
      background: '#ecfeed',
      display: 'flex',
      flexDirection: 'column',
      justifyContent: 'center',
      alignItems: 'center',
    }}
  >
    <Input searchField size='lg' placeholder='SÃ¸k' />
  </div>
)
