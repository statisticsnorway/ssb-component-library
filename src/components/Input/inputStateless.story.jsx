import React from 'react';
import { storiesOf } from '@storybook/react';
import centered from '@storybook/addon-centered/react';
import useState from 'storybook-addon-state';
import Input from './index';

let someValue = '';
const handleChange = e => {
	someValue = e.target.value;
};

const handleSubmit = e => {
	console.log(e.target.value);
};

storiesOf('Input', module).addDecorator(centered)
	.add('with state', () => {
		/**
		 * Just an example to show Input component works fine when the app state
		 * (which supplies the prop value) changes. Here appValue is shared across
		 * two components and updating one reflects in the other
		 */
		const [appValue, setAppValue] = useState('appVal', '');

		const updateValue = e => {
			setAppValue(e.target.value);
		};

		return (
			<div>
				<Input label="orig" value={appValue} handleChange={updateValue} />
				<Input lable="copy" value={appValue} />
			</div>
		);
	})
	.add('Default', () => (
		<div style={{ width: '280px' }}>
			<Input label="Input field" value={someValue} handleChange={handleChange} />
		</div>
	))
	.add('Search field', () => (
		<div style={{ width: '280px' }}>
			<Input ariaLabel="Input field Search" ariaLabelSearchButton="Search" searchField submitCallback={handleSubmit} placeholder="Search text" />
		</div>
	))
	.add('With value', () => (
		<div style={{ width: '280px' }}>
			<Input label="Input field" value="Already filled" />
		</div>
	))
	.add('Disabled', () => (
		<div style={{ width: '280px' }}>
			<Input label="Input field" disabled />
		</div>
	))
	.add('Error', () => (
		<div style={{ width: '280px' }}>
			<Input label="Input field" error errorMessage="Beklager, dette er feil" />
		</div>
	))
	.add('Negative', () => (
		<div style={{
			width: '40em',
			height: '20em',
			background: '#274247',
			display: 'flex',
			flexDirection: 'column',
			justifyContent: 'center',
			alignItems: 'center',
		}}
		>
			<Input label="Input field" negative value={someValue} handleChange={handleChange} />
			<Input label="Input field" negative searchField submitCallback={handleSubmit} placeholder="Search text" />
			<Input label="Input field" negative value={someValue} error errorMessage="Beklager, dette er feil" />
		</div>
	))
	.add('Large input', () => (
		<div style={{
			width: '980px',
			height: '240px',
			background: '#ecfeed',
			display: 'flex',
			flexDirection: 'column',
			justifyContent: 'center',
			alignItems: 'center',
		}}
		>
			<Input searchField size="lg" value="LÃ¸nn" />
		</div>
	));
