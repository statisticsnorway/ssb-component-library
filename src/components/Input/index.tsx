import React, { useState, forwardRef, useEffect } from 'react'
import { Search } from 'react-feather'
import { useId } from '../../utils/useId'
import InputError from '../InputError'
import classNames from '../../utils/utils'

interface InputProps {
  role?: string
  ariaLabelWrapper?: string
  ariaLabel?: string
  ariaLabelledBy?: string
  ariaLabelSearchButton?: string
  name?: string
  className?: string
  disabled?: boolean
  error?: boolean
  errorMessage?: string
  handleChange?: (value: string) => void
  onFocus?: () => void
  onBlur?: () => void
  id?: string
  label?: string
  negative?: boolean
  placeholder?: string
  searchField?: boolean
  size?: string
  submitCallback?: (value: string) => void
  type?: string
  value?: string
}

const Input = forwardRef<HTMLDivElement, InputProps>(
  (
    {
      role,
      ariaLabelWrapper,
      ariaLabel,
      ariaLabelledBy,
      ariaLabelSearchButton = 'search',
      name,
      className = '',
      disabled = false,
      error = false,
      errorMessage,
      handleChange = () => {},
      id,
      label,
      negative = false,
      placeholder,
      searchField = false,
      submitCallback = () => {},
      type = 'text',
      value = '',
      onFocus = () => {},
      onBlur = () => {},
      size,
    },
    ref
  ) => {
    const [inputValue, setInputValue] = useState(value)
    const inputId = id || useId()

    useEffect(() => setInputValue(value), [value])

    const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
      setInputValue(e.target.value)
      handleChange(e.target.value)
    }

    const handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
      if (e.key === 'Enter') {
        submitCallback(inputValue)
      }
    }

    return (
      <div
        className={classNames(
          'ssb-input',
          negative && 'negative',
          error && 'error',
          size === 'lg' && 'input-lg',
          className
        )}
      >
        {label && (
          <label aria-hidden={ariaLabel ? true : undefined} htmlFor={inputId}>
            {label}
          </label>
        )}
        <div ref={ref} className='input-wrapper' role={searchField ? 'search' : role} aria-label={ariaLabelWrapper}>
          <input
            id={inputId}
            name={name}
            disabled={disabled}
            type={type}
            value={inputValue}
            onChange={(e) => handleInputChange(e)}
            onFocus={onFocus}
            onBlur={onBlur}
            placeholder={placeholder}
            aria-label={ariaLabel}
            className={searchField || error ? ' with-icon' : ''}
            onKeyDown={searchField ? (e) => handleKeyDown(e) : undefined}
            aria-describedby={error && errorMessage ? `error_${inputId}` : undefined}
            aria-labelledby={ariaLabelledBy}
          />
          {searchField && (
            <button
              aria-label={ariaLabelSearchButton}
              className='icon-wrapper search-icon'
              onClick={() => submitCallback(inputValue)}
            >
              <Search size={size === 'lg' ? '72' : '18'} />
            </button>
          )}
        </div>
        {error && errorMessage && (
          <InputError negative={negative} errorMessage={errorMessage} id={`error_${inputId}`} />
        )}
      </div>
    )
  }
)

Input.displayName = 'Input'

export default Input
