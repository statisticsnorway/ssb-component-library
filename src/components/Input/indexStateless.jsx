import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import uuid from 'uuid/v4';
import { Search } from 'react-feather';
import InputError from '../InputError';

const Input = ({
	ariaLabel, ariaLabelSearchButton, className, disabled, error, errorMessage, handleChange, id, label, negative, placeholder, searchField, submitCallback, type, value, onFocus, onBlur, size,
}) => {
	const inputId = id || uuid();

	const handleInputChange = e => {
		handleChange(e.target.value);
	};

	const handleKeyDown = e => {
		if (e.key === 'Enter') {
			submitCallback(value);
		}
	};

	return (
		<div className={`ssb-input${negative ? ' negative' : ''}${error ? ' error' : ''}${size === 'lg' ? ' input-lg' : ''}${className ? ` ${className}` : ''}`}>
			{label && <label htmlFor={inputId}>{label}</label>}
			<div className="input-wrapper">
				<input
					id={inputId}
					disabled={disabled}
					type={type}
					value={value}
					onChange={e => handleInputChange(e)}
					onFocus={onFocus}
					onBlur={onBlur}
					placeholder={placeholder}
					aria-label={ariaLabel}
					className={searchField || error ? ' with-icon' : ''}
					onKeyDown={searchField ? e => handleKeyDown(e) : undefined}
				/>
				{searchField && (
					<button aria-label={ariaLabelSearchButton} className="icon-wrapper search-icon" onClick={() => submitCallback(value)}>
						<Search size={size === 'lg' ? '72' : '18'} />
					</button>
				)}
			</div>
			{error && (errorMessage && (
				<InputError negative={negative} errorMessage={errorMessage} />
			))}
		</div>
	);
};

Input.defaultProps = {
	className: '',
	disabled: false,
	error: false,
	handleChange: () => {},
	onFocus: () => {},
	onBlur: () => {},
	negative: false,
	searchField: false,
	submitCallback: () => {},
	type: 'text',
	ariaLabelSearchButton: 'search',
	value: '',
};

Input.propTypes = {
	ariaLabel: PropTypes.string,
	ariaLabelSearchButton: PropTypes.string,
	className: PropTypes.string,
	disabled: PropTypes.bool,
	error: PropTypes.bool,
	errorMessage: PropTypes.string,
	handleChange: PropTypes.func,
	onFocus: PropTypes.func,
	onBlur: PropTypes.func,
	id: PropTypes.string,
	label: PropTypes.string,
	negative: PropTypes.bool,
	placeholder: PropTypes.string,
	searchField: PropTypes.bool,
	size: PropTypes.string,
	submitCallback: PropTypes.func,
	type: PropTypes.string,
	value: PropTypes.string,
};

export default Input;
