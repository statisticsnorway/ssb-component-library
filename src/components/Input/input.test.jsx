import React from 'react';
import {mount, shallow} from 'enzyme';
import Input from './index';
import InputError from '../InputError';

jest.mock('uuid/v4', () => jest.fn(() => 1));

describe('Input component', () => {
	test('Matches the snapshot', () => {
		const wrapper = shallow(<Input>Input</Input>);
		expect(wrapper).toMatchSnapshot();
	});

	test('Renders label if added', () => {
		const wrapper = shallow(<Input label="Label" />);
		expect(wrapper.find('label').exists()).toEqual(true);
	});

	test('Adds optional classes to wrapper', () => {
		const wrapper = shallow(<Input negative error />);
		expect(wrapper.find('.ssb-input').hasClass('negative')).toEqual(true);
		expect(wrapper.find('.ssb-input').hasClass('error')).toEqual(true);
	});

	test('Type is customizable', () => {
		const wrapper = mount(<Input type="email" />);
		expect(wrapper.find('input').props().type).toEqual('email');
	});

	test('Searchable input renders search icon', () => {
		const wrapper = shallow(<Input ariaLabel="Input field Search" searchField placeholder="Search text" />);
		expect(wrapper.find('button').hasClass('search-icon')).toEqual(true);
	});

	test('handleInputChange', () => {
		const handleChange = jest.fn();
		const wrapper = shallow(<Input handleChange={handleChange} />);
		wrapper.find('input').simulate('change', {
			target: { value: 'hello' }
		});
		expect(handleChange).toBeCalledWith('hello');
	});

	test('handle submit', () => {
		const handleSubmit = jest.fn();
		const wrapper = shallow(<Input searchField submitCallback={handleSubmit} />);
		wrapper.find('.icon-wrapper').simulate('click');
		expect(handleSubmit).toBeCalled();
	});

	test('Renders an error message on error', () => {
		const wrapper = shallow(<Input error errorMessage="An error" />);
		expect(wrapper.find(InputError)).toHaveLength(1);
	})

});
