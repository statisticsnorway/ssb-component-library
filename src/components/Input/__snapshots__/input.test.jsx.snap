// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component Adds optional classes to wrapper 1`] = `
<DocumentFragment>
  <div
    class="ssb-input negative error"
  >
    <div
      class="input-wrapper"
    >
      <input
        class=" with-icon"
        id="1"
        type="text"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Input component Matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="ssb-input"
  >
    <div
      class="input-wrapper"
    >
      <input
        class=""
        id="1"
        type="text"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Input component Searchable input renders search icon 1`] = `
<DocumentFragment>
  <div
    class="ssb-input"
  >
    <div
      class="input-wrapper"
      role="search"
    >
      <input
        aria-label="Input field Search"
        class=" with-icon"
        id="1"
        placeholder="Search text"
        type="text"
        value=""
      />
      <button
        aria-label="search"
        class="icon-wrapper search-icon"
      >
        <svg
          fill="none"
          height="18"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="18"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="11"
            cy="11"
            r="8"
          />
          <line
            x1="21"
            x2="16.65"
            y1="21"
            y2="16.65"
          />
        </svg>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
